In [2]:
# importing libraries

In [3]:
# import libraries
import pandas as pd
import numpy as np
import os

In [4]:
# Define the path to the file
file_path = r'C:\Users\User\Downloads\4.3_orders_products\02 Data\Prepared Data\orders_wrangled.csv'

In [5]:
# Import the CSV file into a DataFrame
df_ords = pd.read_csv(file_path)

In [17]:
df = pd.read_csv(r'C:\Users\User\Downloads\4.3_orders_products\products.csv', index_col = False)

In [18]:
df_prods = pd.read_csv(r'C:\Users\User\Downloads\4.3_orders_products\products.csv', index_col=False)

In [9]:
# Create a new data frame
df_test = pd.DataFrame()

In [10]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# Check for mixe types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\User\AppData\Local\Temp\ipykernel_27352\396529503.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [15]:
file_path = r'C:\Users\User\Downloads\4.3_orders_products\02 Data\Original Data\products.csv'

In [19]:
# missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [21]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [22]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [23]:
df_prods.shape

(49693, 5)

In [24]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
# Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [27]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
df_prods_clean.shape

(49677, 5)

In [29]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [31]:
# Run the df.describe() function on your df_ords dataframe
path = 'C:\\Users\\User\\Downloads\\4.3_orders_products\\02 Data\\Prepared Data\\orders_wrangled.csv'

In [32]:
df_ords = pd.read_csv(path)

In [33]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Observation

In [ ]:
Investigating the accuracy of the columns in the df_ords dataframe. After examining the values in each column, no unreasonable values were found. All values in the columns appear to fall within an acceptable and interpretable range for their corresponding columns/variables.

In [56]:
if 'eval_set' in df_ords.columns:
    # Drop the 'eval_set' column permanently
    df_ords = df_ords.drop(columns=['eval_set'], errors='ignore')

In [57]:
# mixed-type data in your df_ords dataframe
(df_ords.head())

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [ ]:
# No mixed data type

In [39]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col) 

C:\Users\User\AppData\Local\Temp\ipykernel_27352\475043532.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\User\AppData\Local\Temp\ipykernel_27352\475043532.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\User\AppData\Local\Temp\ipykernel_27352\475043532.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\User\AppData\Local\Temp\ipykernel_27352\475043532.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\User\AppData\Local\Temp\ipy

In [38]:
#Checking the data types of all the columns in the df_ords dataframe
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


# Observation
There is no mixed-type data found in any of the columns/variables in the df_ords dataframe.

This was confirmed by searching for mixed-type data in a column using the a function and the result was NULL.

The info function also confirmed that no mixed-type data is present.

No mixed type data column/ variable was found.

In [40]:
# missing values in the df_ords dataframe.
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [58]:
df_ords

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [60]:
#Find out the the missing values based on user_id count summarized
df_ords[df_ords['days_since_prior_order'].isnull()].groupby('user_id').count().describe()

order_id  order_number  order_dow  order_hour_of_day  \
count  206209.0      206209.0   206209.0           206209.0   
mean        1.0           1.0        1.0                1.0   
std         0.0           0.0        0.0                0.0   
min         1.0           1.0        1.0                1.0   
25%         1.0           1.0        1.0                1.0   
50%         1.0           1.0        1.0                1.0   
75%         1.0           1.0        1.0                1.0   
max         1.0           1.0        1.0                1.0   

       days_since_prior_order  
count                206209.0  
mean                      0.0  
std                       0.0  
min                       0.0  
25%                       0.0  
50%                       0.0  
75%                       0.0  
max                       0.0

In [81]:
df_ords_clean = df_ords
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True
df_ords_clean

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  new_customer  
0                           NaN          True  
1                          15.0         False  
2                          21.0         False  
3                          29.0         False  
4                          28.0         False  
...                         ...           ...  
3421078                    29.0         False  
3421079                    30.0         False  
3421080                    18.0         False  
3421081                     7.0         False  
3421082                    30.0         False  

[3421083 rows x 7 columns]

# Observation
There are a total of 206209 missing values that were found in the days_since_prior_order column/variable in the original df_ords dataframe.

While the df_ords dataframe has a total of 3421083 rows and 7 columns, the total number of unique user_id is 206209.

This suggests that each unique user_id has one missing value in the days_since_prior_order column, and that missing value belongs to the very first order of each unique user_id.

This makes sense because there is no value/day before each first order.

And for this reason, it is unnecessary to remove these missing values.

No need removing missing values

In [43]:
# duplicates in the df_ords dataframe
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [ ]:
# Observation
no duplicate values found in the df_ords dataframe.

In [59]:
#Chekcing the df_ords dataframe
df_ords

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

# Exporting Data

In [48]:
# Ensure the 'path' variable points to the base directory of your project
path = r'C:\Users\User\Downloads\4.3_orders_products'

In [49]:
# Correctly specify the path for the export
export_path = os.path.join(path, '02-Data', 'Prepared-Data')

In [50]:
# Create the 'Prepared-Data' folder if it does not exist
if not os.path.exists(export_path):
    os.makedirs(export_path)

In [51]:
#Export the cleaned datafrme for df_prods
df_prods_clean_no_dups.to_csv(os.path.join(export_path, 'products_checked.csv'), index=False)